opentreemap/treemap/udf.py

```diff
@@ -1094,43 +1106,270 @@ class UDFModel(UserTrackable, models.Model):
+                    pass
+
+        def iterkeys(self):
+            for k, v in self.iteritems():
+                yield k
+
+        def itervalues(self):
+            for k, v in self.iteritems():
+                yield v
+
+        def items(self):
+            return [i for i in self.iteritems()]
+
+        def keys(self):
+            return [k for k in self.iterkeys()]
+
+        def values(self):
+            return [v for v in self.itervalues()]
+
     def __init__(self, *args, **kwargs):
+        # Model doesn't know anything about `'udfs'`,
+        # so take it out before `super`.
+        udfs_kwarg = kwargs.pop('udfs', {})
+
         super(UDFModel, self).__init__(*args, **kwargs)
-        # Collection UDF audits are handled by the UDFCollectionValue class
+
+        self._setup_udf_model_type()
+
+        hstore_udfs_kwarg = copy.deepcopy(getattr(self, 'hstore_udfs', {}))
+        self._setup_udfs(udfs_kwarg, hstore_udfs_kwarg)
+
+        # Collection UDF audits are handled by the
+        # UserDefinedCollectionValue class
         self._collection_field_names = {udfd.canonical_name
                                         for udfd in self.collection_udfs}
         # This is the whole reason for keeping _do_not_track
         # in addition to the do_not_track class property.
         self._do_not_track |= self.do_not_track | self._collection_field_names
-        self.populate_previous_state()
 
-        self.dirty_collection_udfs = {}
+        # It is the leaf class' responsibility to call
+        # `self.populate_previous_state()` after initialization is
+        # otherwise complete.
+
+        self.dirty_collection_udfs = set()
+
+    def _setup_udf_model_type(self):
+        '''
+        Model type used in searching for UserDefinedFieldDefinition objects
+        related to self.
+
+        Cache it in self.__dict__ for use by __getattr__, to avoid
+        infinite recursion.
+        '''
+        classname = type(self).__name__
+        self._udf_model_type = getattr(self, 'feature_type',
+                                       classname) or classname
+
+    def _setup_udfs(self, udfs={}, hstore_udfs={}):
+        hstore_udfs_kwarg = copy.deepcopy(hstore_udfs)
+        hstore_udfs_kwarg.update(udfs)
+        self.udfs = UDFModel.UdfsProxy(self, 'udfs', 'hstore_udfs',
+                                       **hstore_udfs_kwarg)
 
     @classproperty
     def do_not_track(cls):
-        return UserTrackable.do_not_track | {'udfs'}
+        return UserTrackable.do_not_track | {'udfs', 'hstore_udfs'}
 
     def fields_were_updated(self):
         normal_fields = super(UDFModel, self).fields_were_updated()
 
         return normal_fields or self.dirty_collection_udfs
 
     def get_user_defined_fields(self):
-        if hasattr(self, 'instance'):
-            return udf_defs(self.instance, self._model_name)
-        else:
-            return []
+        # self might not be fully initialized yet. Beware of __getattr__.
```

Comment:
Changed `__getattr__` to call `_get_user_defined_fields_from_dict` directly and forgot to change this comment.

Sometimes `__getattr__` was failing dot notation to get the `self.get_user_defined_fields` method.

Will remove the comment asap.